httpx/auth.py

```diff
@@ -36,3 +40,132 @@ def build_auth_header(self) -> str:
         userpass = b":".join((username, password))
         token = b64encode(userpass).decode().strip()
         return f"Basic {token}"
+
+
+class HTTPDigestAuth:
+
+    ALGORITHM_TO_HASH_FUNCTION = {
+        "MD5": hashlib.md5,
+        "SHA": hashlib.sha1,
+        "SHA-256": hashlib.sha256,
+        "SHA-512": hashlib.sha512,
+    }
+
+    def __init__(
+        self, username: typing.Union[str, bytes], password: typing.Union[str, bytes]
+    ) -> None:
+        self.username = username
+        self.password = password
+
+    def __call__(self, request: AsyncRequest, response) -> AsyncRequest:
+        auth_header = self.build_auth_header(request, response)
+        if auth_header is not None:
+            # TODO: raise instead of returning None?
+            request.headers["Authorization"] = auth_header
+        return request
+
+    def build_auth_header(
+        self, request: AsyncRequest, response: AsyncResponse
+    ) -> typing.Optional[str]:
+        # Retrieve challenge from response header
+        header = response.headers.get("www-authenticate")
+        assert header.lower().startswith("digest")
```

Comment:
We can't assume that "digest" is being offered as the first challenge: https://tools.ietf.org/html/rfc7235#section-4.1

Reply:
I see, thanks. 👍